File: abc.lua
local Proxy = setmetatable({
    __index = function(self, key)
        return self.__self:get(key)
    end;
    __newindex = function(self, key, val)
        return self.__self:set(key, val)
    end;
    __tostring = function(self)
        return self.__self:tostring()
    end;
    __call = function(self, new)
        local node = self.__self
        if not node.value then
            assert(new, ("node '%s' is not defined"):format(node.name))
            node.value = self
            node.index = node.new_index()
        end
        return node.index
    end;
}, {
    __call = function(Proxy, t)
        return setmetatable({__self = t or {}}, Proxy)
    end;
})

local function index_tostring(index)
    if not jit or index < 32767 then
        return tostring(index)
    end
    return ("%d*M+%d"):format(math.floor(index / 32767), index % 32767)
end

local function operand_tostring(t)
    if getmetatable(t) == Proxy then
        local node = t.__self
        assert(node.value, ("node '%s' is not defined"):format(node.name))
        return "x["..index_tostring(node.index).."]"
    else
        return tostring(t)
    end
end

local Node = setmetatable({
    __index = {
        get = function(self, key)
            return self.nodes[key]
        end;
        set = function(self, key, val)
            local proxy = self.nodes[key]
            local node = proxy.__self
            assert(not node.value, ("node '%s' is already defined"):format(node.name))
            local val_type = type(val)
            if val_type == "table" and getmetatable(val) == nil then
                for k, v in pairs(val) do
                    proxy[k] = v
                end
                return
            end
            assert(val_type == "table", ("it is forbidden to assign a %s"):format(val_type))
            node.value = val
            node.index = self.new_index()
        end;
        tostring = function(self)
            local t = {}
            if next(self.nodes) then
                for _, v in pairs(self.nodes) do
                    t[#t+1] = tostring(v)
                end
            else
                assert(self.value, ("node '%s' is not defined"):format(self.name))
            end
            if self.value then
                t[#t+1] = "y["..index_tostring(self.index).."]="..operand_tostring(self.value).." -- "..self.name
            end
            return table.concat(t, "\n")
        end;
    };
}, {
    __call = function(Node, name, indexer)
        assert(name, "name required")
        assert(indexer, "indexer required")
        local t = setmetatable({
            name = name;
            index = 0;
            value = false;
            nodes = setmetatable({}, {
                __index = function(self, key)
                    local node = Node(name.."."..key, indexer)
                    self[key] = node
                    return node
                end;
            });
            new_index = indexer;
        }, Node)
        return Proxy(t)
    end;
})

local Or = {
    __tostring = function(self)
        if _VERSION < "Lua 5.3" then
            return "O("..operand_tostring(self.lhs)..","..operand_tostring(self.rhs)..")"
        else
            return "("..operand_tostring(self.lhs).."|"..operand_tostring(self.rhs)..")"
        end
    end;
}
local Xor = {
    __tostring = function(self)
        if _VERSION < "Lua 5.3" then
            return "X("..operand_tostring(self.lhs)..","..operand_tostring(self.rhs)..")"
        else
            return "("..operand_tostring(self.lhs).."~"..operand_tostring(self.rhs)..")"
        end
    end;
}
local And = {
    __tostring = function(self)
        if _VERSION < "Lua 5.3" then
            return "A("..operand_tostring(self.lhs)..","..operand_tostring(self.rhs)..")"
        else
            return "("..operand_tostring(self.lhs).."&"..operand_tostring(self.rhs)..")"
        end
    end;
}
local Not = {
    __tostring = function(self)
        if _VERSION < "Lua 5.3" then
            return "N("..operand_tostring(self.rhs)..")"
        else
            return "~("..operand_tostring(self.rhs)..")"
        end
    end;
}

local function check_operand(operand)
    local mt = getmetatable(operand)
    assert(mt == Proxy or mt == Or or mt == And or mt == Not or mt == Xor, "unknown type")
end

local bor = function(self, other)
    check_operand(other)
    return setmetatable({op = "bor", lhs = self, rhs = other}, Or)
end

local bxor = function(self, other)
    check_operand(other)
    return setmetatable({op = "bxor", lhs = self, rhs = other}, Xor)
end

local band = function(self, other)
    check_operand(other)
    return setmetatable({op = "band", lhs = self, rhs = other}, And)
end

local bnot = function(self)
    return setmetatable({op = "bnot", rhs = self}, Not)
end

Proxy.__add = bor
Proxy.__sub = bxor
Proxy.__mul = band
Proxy.__unm = bnot

Proxy.__bor = bor
Proxy.__bxor = bxor
Proxy.__band = band
Proxy.__bnot = bnot

Or.__add = bor
Or.__sub = bxor
Or.__mul = band
Or.__unm = bnot

Or.__bor = bor
Or.__bxor = bxor
Or.__band = band
Or.__bnot = bnot

And.__add = bor
And.__sub = bxor
And.__mul = band
And.__unm = bnot

And.__bor = bor
And.__bxor = bxor
And.__band = band
And.__bnot = bnot

Not.__add = bor
Not.__sub = bxor
Not.__mul = band
Not.__unm = bnot

Not.__bor = bor
Not.__bxor = bxor
Not.__band = band
Not.__bnot = bnot

local function Model(indexer)
    local index = 0
    indexer = indexer or function()
        index = index + 1
        return index
    end
    return Node("", indexer)
end

local function Build(model)
    local self = model.__self
    if self.index == 0 then
        self.index = self.new_index() - 1
    end
    local len = self.index
    local src = ([[
local bit = bit or bit32
local O, A, N
if _VERSION < "Lua 5.3" then
O, X, A, N = bit.bor, bit.bxor, bit.band, bit.bnot
end
local mt = {
    __tostring = function(self)
        local t = {}
        for i = 1, #self do
            t[#t+1] = tostring(self[i])
        end
        return "["..table.concat(t, ", ").."]"
    end;
}
local x = setmetatable({}, mt)
local y = setmetatable({}, mt)
for i = 1, %d do x[i] = 0; y[i] = 0 end
local function tick()
    local M = 32767
%s
    x, y = y, x
    return x
end
return tick, x
]]):format(len, tostring(model))
    local f, err = load(src)
    assert(f, err)
    local tick, state = f()
    return tick, state, src
end

return {
    Model = Model;
    Build = Build;
}

File: abc.py

LOAD = 1
AND = 2
OR = 3
XOR = 4
NOT = 5
STORE = 6

class Base:
    def __invert__(self):
        return Not(self)

    def __and__(self, other):
        if not isinstance(other, Base):
            raise Exception(f"unknown type: {other}")
        return And(self, other)

    def __or__(self, other):
        if not isinstance(other, Base):
            raise Exception(f"unknown type: {other}")
        return Or(self, other)

    def __xor__(self, other):
        if not isinstance(other, Base):
            raise Exception(f"unknown type: {other}")
        return Xor(self, other)

def emit_operand(operand, code):
    if isinstance(operand, Model):
        code.append(LOAD)
        code.extend(operand().to_bytes(4, byteorder='little'))
    else:
        operand.emit(code)

class Not(Base):
    def __init__(self, rhs):
        self.rhs = rhs

    def emit(self, code):
        emit_operand(self.rhs, code)
        code.append(NOT)

class And(Base):
    def __init__(self, lhs, rhs):
        self.lhs = lhs
        self.rhs = rhs

    def emit(self, code):
        emit_operand(self.lhs, code)
        emit_operand(self.rhs, code)
        code.append(AND)

class Or(Base):
    def __init__(self, lhs, rhs):
        self.lhs = lhs
        self.rhs = rhs

    def emit(self, code):
        emit_operand(self.lhs, code)
        emit_operand(self.rhs, code)
        code.append(OR)

class Xor(Base):
    def __init__(self, lhs, rhs):
        self.lhs = lhs
        self.rhs = rhs

    def emit(self, code):
        emit_operand(self.lhs, code)
        emit_operand(self.rhs, code)
        code.append(XOR)

class Model(Base):
    def __init__(self, name="", indexer=None):
        self.__dict__["_Model__name"] = name
        self.__dict__["_Model__value"] = None
        self.__dict__["_Model__index"] = None
        self.__dict__["_Model__nodes"] = {}
        if indexer is None:
            index = -1
            def indexer(current=False):
                nonlocal index
                if current:
                    return index + 1
                index += 1
                return index
        self.__dict__["_Model__indexer"] = indexer

    def __call__(self, new=False):
        if self.__index is None:
            if not new:
                raise Exception(f"node '{self.__name}' is not defined")
            self.__dict__["_Model__value"] = self
            self.__dict__["_Model__index"] = self.__indexer()
        return self.__index

    def __getitem__(self, key):
        node = self.__nodes.get(key)
        if node is None:
            node = Model(f"{self.__name}.{key}", self.__indexer)
            self.__nodes[key] = node
        return node

    def __getattr__(self, name):
        return self.__getitem__(name)

    def __setitem__(self, key, value):
        node = self.__getitem__(key)
        if node._Model__value is not None:
            raise Exception(f"node '{node.__name}' is already defined")
        if isinstance(value, dict):
            for k, v in value.items():
                node[k] = v
            return
        if not isinstance(value, Base):
            raise Exception(f"it is forbidden to assign a {value}")
        node.__dict__["_Model__value"] = value
        node.__dict__["_Model__index"] = self.__indexer()

    def __setattr__(self, name, value):
        self.__setitem__(name, value)

    def __emit(self, code):
        nodes = self.__nodes
        if len(nodes) > 0:
            for key in nodes:
                nodes[key].__emit(code)
            if self.__value is None:
                return
        else:
            if self.__value is None:
                raise Exception(f"node '{self.__name}' is not defined")
        emit_operand(self.__value, code)
        code.append(STORE)
        code.extend(self().to_bytes(4, byteorder='little'))

def Compile(model):
    size = model._Model__indexer(True)
    code = bytearray()
    code.extend(size.to_bytes(4, byteorder='little'))
    model._Model__emit(code)
    return code

def Build(model):
    code = Compile(model)
    size = (code[3] << 24) + (code[2] << 16) + (code[1] << 8) + code[0]
    import numpy as np
    from numba import jit, uint64
    prg = np.array(code, np.uint8)
    stack = np.zeros(1000, np.uint64)
    state = np.zeros(size, np.uint64)
    _state = np.zeros(size, np.uint64)
    @jit((uint64[:], uint64[:], uint64[:]), nopython=True, nogil=True)
    def tickjit(state, _state, stack):
        i = 0
        ip = 0
        sp = -1
        while i < size:
            ip += 1
            b = prg[ip]
            if b == LOAD:
                sp += 1
                ip += 4
                idx = (prg[ip] << 24) + (prg[ip-1] << 16) + (prg[ip-2] << 8) + prg[ip-3]
                stack[sp] = state[idx]
            elif b == AND:
                sp -= 1
                stack[sp] = stack[sp] & stack[sp+1]
            elif b == OR:
                sp -= 1
                stack[sp] = stack[sp] | stack[sp+1]
            elif b == XOR:
                sp -= 1
                stack[sp] = stack[sp] ^ stack[sp+1]
            elif b == NOT:
                stack[sp] = ~stack[sp]
            elif b == STORE:
                ip += 4
                idx = (prg[ip] << 24) + (prg[ip-1] << 16) + (prg[ip-2] << 8) + prg[ip-3]
                _state[idx] = stack[sp]
                sp -= 1
                i += 1

    def tick():
        nonlocal state, _state, stack
        tickjit(state, _state, stack)
        state, _state = _state, state
        return state
    return tick

m = Model()

quacks = m.quacks(True)
flies = m.flies(True)
swims = m.swims(True)
croaks = m.croaks(True)

m.duck = m.quacks & (m.flies | m.swims)
m.frog = m.croaks & m.swims & ~m.flies

tick = Build(m)

state = tick()

print(state)

state[quacks] = 0
state[flies] = 0
state[swims] = 1
state[croaks] = 1

print(state)

state = tick()
print(state)

print("is it a duck?", state[m.duck()])
print("is it a frog?", state[m.frog()])

########

# import time

# start = time.time()

# it = Model()

# it.quacks = it.quacks
# it.flies = it.flies
# it.swims = it.swims
# it.croaks = it.croaks

# for i in range(1_000_000):
#     it.duck[i] = it.quacks & (it.flies | it.swims)

########

# code = Compile(it)

# with open("code.data", "wb") as f:
#     f.write(code)

########

# tick = Build(it)

# print("compilation time: ", time.time() - start)

# tick()

# start = time.time()

# tick()

# print("calculation time: ", time.time()-start)